walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,151 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
 
 public final class FindMeetingQuery {
+  // Return the possible times at which the meeting be set on the day.
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+
+    // All the possible times to arrange a meeting on the day.
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    // All the people who are required to attend the meeting. 
+    Collection<String> requestAttendees = request.getAttendees();
+
+    // All the people whose attendance is optional. 
+    Collection<String> optionalAttendees = request.getOptionalAttendees();
+
+    // Events with required attendees. We assume events are always added in a chronological order. 
+    List<Event> requiredEvents = new ArrayList<>();
+
+    // Events with optional attendees. We assume events are always added in a chronological order. 
+    List<Event> optionalEvents = new ArrayList<>();
+
+    if (events.isEmpty() || (requestAttendees.isEmpty() && optionalAttendees.isEmpty())) {
+      // If the request has a meeting's duration longer than a day, there are no possible options.
+      if (request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+        return Arrays.asList();
+      }
+
+      // If there are no events or no required attendees and the meeting duration is within a day, the whole day is available to set up a meeting. 
+      else {
+        return Arrays.asList(TimeRange.WHOLE_DAY);
+      }
+    }
+
+    for (Event event : events) {
+      for (String requiredAttendee : requestAttendees) {
+        if (event.getAttendees().contains(requiredAttendee)) {
+          if (!requiredEvents.contains(event)) {
+            requiredEvents.add(event);
+          }
+        }
+      }
+
+      if (!optionalAttendees.isEmpty()) {
+        for (String optionalAttendee : optionalAttendees) {
+          if (event.getAttendees().contains(optionalAttendee)) {
+            if (!optionalEvents.contains(event)) {
+              optionalEvents.add(event);
+            }
+          }
+        }
+      }
+    }
+
+    if (requiredEvents.size() > 0 || (requiredEvents.isEmpty() && !optionalEvents.isEmpty())) {
+      // If there is only 1 event, the possible times are before and after the event. 
+      if (requiredEvents.size() == 1 || (requiredEvents.isEmpty() && optionalEvents.size() == 1)) {
```

Comment:
There are a lot of nested if/else blocks here -- is it possible to simplify the logic here to make it clearer what's happening? One option might be to separate out some of these possibilities and return early

Reply:
For example, the outer loop could be replaced with an initial "if (requiredEvents.isEmpty() && optionalEvents.isEmpty()) { return Arrays.asList(TimeRange.WHOLE_DAY); }"

Without a subsequent 'else' because we've already returned from that loop.